backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/AbstractElasticsearchSearchNestedPredicateBuilder.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.elasticsearch.search.predicate.impl;
+
+import java.util.LinkedList;
+import java.util.List;
+
+import org.hibernate.search.backend.elasticsearch.gson.impl.JsonAccessor;
+
+import com.google.gson.JsonObject;
+
+public abstract class AbstractElasticsearchSearchNestedPredicateBuilder extends AbstractElasticsearchSearchPredicateBuilder {
+
+	private static final JsonAccessor<String> PATH_ACCESSOR = JsonAccessor.root().property( "path" ).asString();
+	private static final JsonAccessor<JsonObject> QUERY_ACCESSOR = JsonAccessor.root().property( "query" ).asObject();
+
+	private final List<String> nestedPathHierarchy;
+
+	public AbstractElasticsearchSearchNestedPredicateBuilder(List<String> nestedPathHierarchy) {
+		this.nestedPathHierarchy = nestedPathHierarchy;
+	}
+
+	@Override
+	public final JsonObject build(ElasticsearchSearchPredicateContext context) {
+		JsonObject result = super.build( context );
+		if ( nestedPathHierarchy != null && !nestedPathHierarchy.isEmpty() ) {
+			result = applyImplicitNested( result, new LinkedList<>( nestedPathHierarchy ), context );
+		}
+		return result;
+	}
+
+	private JsonObject applyImplicitNested(JsonObject partialResult, List<String> nestedPathHierarchy,
+			ElasticsearchSearchPredicateContext context) {
+		if ( nestedPathHierarchy.isEmpty() ) {
+			return partialResult;
+		}
+
+		String lastPath = nestedPathHierarchy.remove( 0 );
```

Comment:
This line and the ones below could be converted to a simple loop instead of using recursion + a mutable collection. It would probably be easier to read, and there would be one fewer LinkedList allocation.

E.g.:


```
	JsonObject result = partialResult;
	for ( String nestedPath : nestedPathHierarchy ) {
		if ( context.isNested( nestedPath ) ) {
			// skip this implicit nested
			continue;
		}
    
		JsonObject innerObject = new JsonObject();

		PATH_ACCESSOR.set( innerObject, lastPath );
		QUERY_ACCESSOR.set( innerObject, applyImplicitNested( partialResult, nestedPathHierarchy, context ) );

		JsonObject outerObject = new JsonObject();
		outerObject.add( "nested", innerObject );

		result = outerObject; 
	}
```

Reply:
The example runs `applyImplicitNested( partialResult, nestedPathHierarchy, context ) ` inside the loop. Of course it always possible to convert a recursion with a loop but I don't think that the resulting code would be simplified by that. Do you?

Reply:
Maybe traversing the `nestedPathHierarchy` in the inverse order would work. I'm trying now...

Reply:
> Do you?

Yes. If I compare this recursion solution with what you did in `org.hibernate.search.backend.elasticsearch.search.sort.impl.AbstractElasticsearchSearchNestedSortBuilder#enrichInnerObject`, I must say that other looping solution is much clearer.

I suppose the problem is more complex, though... I was just hoping there would be a clearer solution :)

Reply:
It seems to be working:

``` java
private static JsonObject applyImplicitNestedStepIt(JsonObject partialResult, List<String> nestedPathHierarchy, ElasticsearchSearchPredicateContext context) {
	JsonObject result = partialResult;

	// traversing the nestedPathHierarchy in the inverted order
	for ( int i = 0; i < nestedPathHierarchy.size(); i++ ) {
		String path = nestedPathHierarchy.get( nestedPathHierarchy.size() - 1 - i );
		if ( context.isExplicitNested( path ) ) {
			// skip this implicit nested
			continue;
		}

		JsonObject innerObject = new JsonObject();

		PATH_ACCESSOR.set( innerObject, path );
		QUERY_ACCESSOR.set( innerObject, result );

		JsonObject outerObject = new JsonObject();
		outerObject.add( "nested", innerObject );
		result = outerObject;
	}

	return result;
}
```

Reply:
The trick is the inverted traversing order... So that we can reuse the nested JSON object